### ISSUE TITLE
Feature Suggestion: Fittable Skip-conditions & Pre-conditions

### ISSUE THREAD

**Issue by shaypal5** [pdpipe/pdpipe#20]
Fittable conditions should be defined to be any callable that given a `pandas.DataFrame` object returns `True` or `False`. A type, however, should be defined, that can have a state of being fitted. On their first call they check the conditions they represent and return the appropriate boolean value, but they also save it, returning it on every future call, instead of checking for the condition again.

However, these condition objects should also have a `fittable` constructor parameter, that if supplied with `False` prevents them from being fitted.

Then, pipeline stages can be supplied with callables to two new constructor parameters:
1. `prec` - Sets the given callable as the pre-condition for the pipeline stage, instead of the built-in `_prec` method. Should be provided with unfittable conditions, to have meaningful behaviour (but any behaviour is supported).
2. `skip` - Sets the given callable as the condition for skipping the pipeline stage. Probably more use here for fittable conditions, so a stage skipped in `fit_transform` is always skipped in transform, but I can see how unfittable conditions might also be useful.

I like this design because it delegates these behaviours to the condition objects themselves instead of deciding it for "everyone" in the design of the `PipelineStage` object itself. The main downside I see is we can't get a nice behaviour like supplying skip with a simple callable, like a lambda, and just wrap it a fittable condition object, because now we can't assume skip-conditions should not be checked in transform time, but I think it's a small price to pay.

This conditions should definitely play nice and use the fittable column qualifiers suggested in #17 (which should be generat-able through the `pdpipe.cq` factory sub-module, or something of the like).

I think I should also write a factory sub-module for these conditions, so easy generation of logic can be done with:
```python
pdp.PCA(
  columns=pdp.cq.Dtype(np.int64),
  skip=pdp.cond.HasColumnsOfDtype(np.int64),
)
```

Finally, this all makes ad-hoc stages (made by instantiating `AdHocPipelineStage` objects) much nicer, because their `prec` parameter can now be populated with this conditions.

**Comment by shaypal5** [comment]
Implemented by PR #26 ! :)

**Comment by shaypal5** [comment]
Released in `v0.0.46`: https://github.com/pdpipe/pdpipe/releases/tag/v0.0.46
